src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs

```diff
@@ -0,0 +1,2164 @@
+                        value,
+                        string.Format(SR.net_http_value_must_be_greater_than, 0));
+                }
+
+                CheckDisposedOrStarted();
+                _maxResponseDrainSize = value;
+            }
+        }
+        #endregion
+
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing && !_disposed)
+            {
+                _disposed = true;
+
+                if (_sessionHandle != null)
+                {
+                    _sessionHandle.DangerousRelease();
+                    SafeInternetHandle.DisposeAndClearHandle(ref _sessionHandle);
+                }
+            }
+
+            base.Dispose(disposing);
+        }
+
+#if HTTP_DLL
+        protected internal override Task<HttpResponseMessage> SendAsync(
+            HttpRequestMessage request,
+            CancellationToken cancellationToken)
+#else
+        protected override Task<HttpResponseMessage> SendAsync(
+            HttpRequestMessage request,
+            CancellationToken cancellationToken)
+#endif
+        {
+            if (request == null)
+            {
+                throw new ArgumentNullException("request", SR.net_http_handler_norequest);
+            }
+
+            // Check for invalid combinations of properties.
+            if (_proxy != null && _windowsProxyUsePolicy != WindowsProxyUsePolicy.UseCustomProxy)
+            {
+                throw new InvalidOperationException(SR.net_http_invalid_proxyusepolicy);
+            }
+
+            if (_windowsProxyUsePolicy == WindowsProxyUsePolicy.UseCustomProxy && _proxy == null)
+            {
+                throw new InvalidOperationException(SR.net_http_invalid_proxy);
+            }
+
+            if (_cookieUsePolicy == CookieUsePolicy.UseSpecifiedCookieContainer &&
+                _cookieContainer == null)
+            {
+                throw new InvalidOperationException(SR.net_http_invalid_cookiecontainer);
+            }
+
+            CheckDisposed();
+
+            SetOperationStarted();
+
+            TaskCompletionSource<HttpResponseMessage> tcs = new TaskCompletionSource<HttpResponseMessage>();
+
+            // Create RequestState object and save current values of handler settings.
+            RequestState state = new RequestState();
+            state.Tcs = tcs;
+            state.CancellationToken = cancellationToken;
+            state.RequestMessage = request;
+            state.Handler = this;
+            state.CheckCertificateRevocationList = _checkCertificateRevocationList;
+            state.ServerCertificateValidationCallback = _serverCertificateValidationCallback;
+            state.WindowsProxyUsePolicy = this.WindowsProxyUsePolicy;
+            state.Proxy = this.Proxy;
+            state.ServerCredentials = _serverCredentials;
+
+            try
+            {
+                Task.Factory.StartNew(StartRequest, state);
+            }
+            catch (Exception ex)
```

Comment:
@stephentoub To be honest, I wasn't sure. :) But it was important to capture the `Exception` and  put it inside an `HttpRequestException` since that is the design contract for the API.  I'm open to suggestions on what the best pattern is for trapping exceptions from `Task.Factory.StartNew`